src/js/player.js

```diff
@@ -1623,37 +1624,51 @@ class Player extends Component {
   }
 
   /**
-   * start media playback
+   * Attempt to begin playback at the first opportunity.
    *
    * @return {Promise|undefined}
-   *         Returns a `Promise` if the browser returns one, for most browsers this will
-   *         return undefined.
+   *         Returns a `Promise` only if the browser returns one and the player
+   *         is ready to begin playback. For some browsers and all non-ready
+   *         situations, this will return `undefined`.
    */
   play() {
-    if (this.changingSrc_) {
-      this.ready(function() {
-        const retval = this.techGet_('play');
 
-        // silence errors (unhandled promise from play)
-        if (retval !== undefined && typeof retval.then === 'function') {
-          retval.then(null, (e) => {});
-        }
+    // If the tech is not ready, queue up another call to `play()` for when it
+    // is. This will loop back into this method for another attempt at playback
+    // when the tech is ready.
+    if (!this.isReady_) {
+
+      // Bail out if we're already waiting for `ready`!
+      if (this.playWaitingForReady_) {
+        return;
+      }
+
+      this.playWaitingForReady_ = true;
+      this.ready(() => {
+        this.playWaitingForReady_ = false;
+        silencePromise(this.play());
       });
 
-    // Only calls the tech's play if we already have a src loaded
-    } else if (this.isReady_ && (this.src() || this.currentSrc())) {
-      return this.techGet_('play');
-    } else {
-      this.ready(function() {
-        this.tech_.one('loadstart', function() {
-          const retval = this.play();
+    // If the tech is ready, but we are currently changing sources, wait for
+    // the next "loadstart" on the tech. Again, this will loop back into this
+    // method for another attempt at playback when we are no longer changing
+    // sources.
+    } else if (this.changingSrc_) {
 
-          // silence errors (unhandled promise from play)
-          if (retval !== undefined && typeof retval.then === 'function') {
-            retval.then(null, (e) => {});
-          }
-        });
+      // Bail out if we're already waiting for `loadstart`!
+      if (this.playWaitingForLoadstart_) {
+        return;
+      }
+
+      this.playWaitingForLoadstart_ = true;
+      this.tech_.one('loadstart', () => {
+        this.playWaitingForLoadstart_ = false;
+        silencePromise(this.play());
       });
+
+    // Finally, if we have a source, we can attempt playback for real.
+    } else if (this.src() || this.currentSrc()) {
+      return this.techGet_('play');
     }
```

Comment:
is it possible that we end up in an `else` condition?

Reply:
I've been trying to think of one that isn't captured by the previous two statements, but I haven't been able to. It probably makes sense to reverse this conditional block so the `ready` call comes in the `else` block.

Reply:
Actually, yeah, the only condition that we'd have that would be an else would be if the tech is ready, we're not changing sources, and we don't have a source.